### PR TITLE
3218 put mapping operation for incremental index mapping updates inside search

### PR DESCRIPTION
**Pull request by fractal3000**
Fixes: #3218

### CODE REVIEW

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+    protected ComparingState isIndexSettingsActual(IndexConfiguration indexConfiguration, GetIndexResponse indexResponse) {
+        Map<String, Settings> settings = indexResponse.getSettings();
+        Settings currentSettings = settings.get(indexConfiguration.getIndexName());
+        Settings actualSettings = indexConfiguration.getSettings();
+        return settingsComparator.compare(currentSettings, actualSettings);
+    }
+
+    public class ConfiguarionComparingResult {
+        private final boolean compartible;
```
Comment (Gavrilov-Ivan) [inline jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java:80]: compatible?

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        private final boolean mappingMustBeActualized;
```
Comment (Gavrilov-Ivan) [inline jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java:81]: Confusing: 'mappingMustBeActualized' - is it required to recreate index or support update (so it's not 'must', but 'can')? If this point also required to take into account the 'compatible' value - it's hard-to-understand and allow to create result with impossible values combination.

Need to simplify

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/SearchMappingComparator.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+        return result;
+    }
+
+    private ComparingState compareDataTypes(String type, String type1) {
+        return type.equals(type1) ? ComparingState.EQUAL: ComparingState.NOT_COMPATIBLE;
+    }
+
+    private boolean isCorrectLeafElement(Map<String, Object> mapping) {
+        return mapping.size() == 1 && mapping.get(TYPE_KEY) != null && mapping.get(TYPE_KEY) instanceof String;
```
Comment (Gavrilov-Ivan) [inline jmix-search/search/src/main/java/io/jmix/search/index/impl/SearchMappingComparator.java:66]: Didn't get this, tbh

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/SearchSettingsComparator.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.jmix.search.index.impl;
+
+import org.elasticsearch.common.settings.Settings;
+import org.springframework.stereotype.Component;
+
+@Component("zndfl_SearchSettingsComparator")
```
Comment (Gavrilov-Ivan) [inline jmix-search/search/src/main/java/io/jmix/search/index/impl/SearchSettingsComparator.java:22]: zndfl?

**Review by Gavrilov-Ivan — CHANGES_REQUESTED** [review]

File: jmix-search/search/src/main/java/io/jmix/search/index/ESIndexManager.java [hunk 5]
```diff
@@ -133,4 +133,11 @@ public interface ESIndexManager {
      * @return {@link IndexSynchronizationStatus}
      */
     IndexSynchronizationStatus synchronizeIndexSchema(IndexConfiguration indexConfiguration);
+
+    /**
+     * TODO
+     * @param indexConfiguration
+     * @return
+     */
+    boolean saveIndexMapping(IndexConfiguration indexConfiguration);
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/ESIndexManager.java]: Поправил

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/ComparingState.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.jmix.search.index.impl;
+
+public enum ComparingState {
+    EQUAL(true, false),
+    COMPATIBLE(true, true),
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/ComparingState.java:21]: Как вариант, можно не поскупиться и назвать NOT_EQUAL_BUT_COMPATIBLE

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/ESIndexManagerImpl.java [hunk 7]
```diff
@@ -172,7 +168,8 @@ public IndexValidationStatus validateIndex(IndexConfiguration indexConfiguration
 
         IndexValidationStatus status;
         if (isIndexExist(indexConfiguration.getIndexName())) {
-            if (isIndexActual(indexConfiguration)) {
+            IndexConfigurationsChecker.ConfiguarionComparingResult result = isIndexActual(indexConfiguration);
+            if (result.isCompartible()) {
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/ESIndexManagerImpl.java]: Fixed

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/ComparingState.java [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    NOT_COMPATIBLE(false, false);
+
+    private final boolean reindexingNotRequired;
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/ComparingState.java:24]: Why?

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+import org.springframework.stereotype.Component;
+
+import java.util.Map;
+
+@Component
+public class IndexConfigurationsChecker {
+    private final SearchMappingComparator mappingComparator;
+    private final SearchSettingsComparator settingsComparator;
+
+    public IndexConfigurationsChecker(SearchMappingComparator searchMappingChecker, SearchSettingsComparator settingsComparator) {
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java]: Agree. Fixed

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java [hunk 10]
```diff
@@ -0,0 +1,102 @@
...
+import org.elasticsearch.common.settings.Settings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.stereotype.Component;
+
+import java.util.Map;
+
+@Component
+public class IndexConfigurationsChecker {
+    private final SearchMappingComparator mappingComparator;
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java]: May be IndexMappingComparator and IndexSettingsComparator?

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java (hunk 1 above)
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java]: Fixed. And class name corrected.

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/SearchSettingsComparator.java (hunk 4 above)
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/SearchSettingsComparator.java:22]: Fixed. And component names to other classes were added.

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/ComparingState.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/ComparingState.java:21]: Учёл в новом варианте реализации

[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/ComparingState.java:21]: Took into account in new option of implementation.

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/ComparingState.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/ComparingState.java:24]: Agree

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java (hunk 10 above)
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java]: Fixed

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/MappingFieldComparator.java [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+@Component("search_MappingFieldComparator")
+public class MappingFieldComparator {
+    public static final String TYPE_KEY = "type";
+
+    public boolean isLeafField(@NotNull Map<String, Object> mapping) {
+        return mapping.get(TYPE_KEY) != null && mapping.get(TYPE_KEY) instanceof String;
+    }
+
+    public IndexMappingComparator.MappingComparingResult compareLeafFields(@NotNull Map<String, Object> searchIndexMapping, @NotNull Map<String, Object> applicationMapping) {
+        if (isNull(searchIndexMapping) || isNull(applicationMapping)) throw new NullPointerException();
```
Comment (Gavrilov-Ivan) [inline jmix-search/search/src/main/java/io/jmix/search/index/impl/MappingFieldComparator.java:35]: Preconditions

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/StartupIndexSynchronizer.java [hunk 12]
```diff
@@ -65,6 +65,7 @@ protected void postConstruct() {
                         status);
                 switch (status) {
                     case CREATED:
+                    case INDEX_MAPPING_UPDATED:
```
Comment (Gavrilov-Ivan) [inline jmix-search/search/src/main/java/io/jmix/search/index/impl/StartupIndexSynchronizer.java:68]: available.add(config);

**Review by Gavrilov-Ivan — CHANGES_REQUESTED** [review]

**Review by Gavrilov-Ivan — CHANGES_REQUESTED** [review]

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/ESIndexManagerImpl.java [hunk 13]
```diff
@@ -227,16 +224,49 @@ public IndexSynchronizationStatus synchronizeIndexSchema(IndexConfiguration inde
...
 
+    @Override
+    public boolean saveIndexMapping(IndexConfiguration indexConfiguration) {
+        PutMappingRequest request = new PutMappingRequest(indexConfiguration.getIndexName());
+
+        String mappingBody;
+        try {
+            mappingBody = objectMapper.writeValueAsString(indexConfiguration.getMapping());
+        } catch (JsonProcessingException e) {
+            throw new RuntimeException("Unable to create index '" + indexConfiguration.getIndexName() + "': Failed to parse index definition", e);
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/ESIndexManagerImpl.java]: Fixed

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java (hunk 2 above)
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexConfigurationsChecker.java]: Fixed

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/SearchMappingComparator.java (hunk 3 above)
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/SearchMappingComparator.java]: Fixed

File: jmix-search/search/src/main/java/io/jmix/search/index/ESIndexManager.java [hunk 14]
```diff
@@ -133,4 +133,11 @@ public interface ESIndexManager {
      * @return {@link IndexSynchronizationStatus}
      */
     IndexSynchronizationStatus synchronizeIndexSchema(IndexConfiguration indexConfiguration);
+
+    /**
+     * TODO
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/ESIndexManager.java:138]: fixed and method signature was corrected

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexMappingComparator.java [hunk 15]
```diff
@@ -0,0 +1,81 @@
...
+
+        if (result == MappingComparingResult.MAPPINGS_ARE_EQUAL && applicationMapping.size() > searchIndexMapping.size()) {
+            return MappingComparingResult.INDEX_MAPPING_CAN_BE_UPDATED;
+        }
+
+        return result;
+    }
+
+    enum MappingComparingResult implements ConfigurationPartComparingResult{
+        MAPPINGS_ARE_EQUAL,
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/IndexMappingComparator.java:67]: fixed

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/ESIndexManagerImpl.java [hunk 16]
```diff
@@ -172,7 +168,8 @@ public IndexValidationStatus validateIndex(IndexConfiguration indexConfiguration
 
         IndexValidationStatus status;
         if (isIndexExist(indexConfiguration.getIndexName())) {
-            if (isIndexActual(indexConfiguration)) {
+            IndexConfigurationComparator.ConfigurationComparingResult result = compareWithIndexConfiguration(indexConfiguration);
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/ESIndexManagerImpl.java:171]: fixed

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/ESIndexManagerImpl.java [hunk 17]
```diff
@@ -227,16 +224,42 @@ public IndexSynchronizationStatus synchronizeIndexSchema(IndexConfiguration inde
...
                 indexConfiguration.getIndexName(), indexConfiguration.getEntityName(), strategy);
         IndexSynchronizationStatus status;
         boolean indexExist = isIndexExist(indexConfiguration.getIndexName());
         if (indexExist) {
-            boolean indexActual = isIndexActual(indexConfiguration);
-            if (indexActual) {
-                status = IndexSynchronizationStatus.ACTUAL;
-                indexStateRegistry.markIndexAsAvailable(indexConfiguration.getEntityName());
+            IndexConfigurationComparator.ConfigurationComparingResult result = compareWithIndexConfiguration(indexConfiguration);
+            if (!result.indexRecreatingIsRequired()) {
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/ESIndexManagerImpl.java:255]: Fixed

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/ESIndexManagerImpl.java [hunk 18]
```diff
@@ -248,6 +271,28 @@ protected IndexSynchronizationStatus synchronizeIndexSchema(IndexConfiguration i
         return status;
     }
 
+    private IndexSynchronizationStatus updateIndexConfiguration(IndexConfiguration indexConfiguration, IndexSchemaManagementStrategy strategy, IndexConfigurationComparator.ConfigurationComparingResult result) {
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/ESIndexManagerImpl.java:274]: Иван, сделаю protected. Но вообще, делать все методы protected- это риск что при переопределении метода, результат работы других методов, которые его используют будет непредсказуемым.

[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/ESIndexManagerImpl.java:274]: Это нарушает принцип открытости-закрытости.

[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/ESIndexManagerImpl.java:274]: Fixed, but practice to make all API methods protected conflicts with "Open-close" principle and leads to unexpected behaviour after overriding such methods.

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/MappingFieldComparator.java (hunk 11 above)
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/MappingFieldComparator.java:35]: Fixed

File: jmix-search/search/src/main/java/io/jmix/search/index/impl/StartupIndexSynchronizer.java (hunk 12 above)
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/impl/StartupIndexSynchronizer.java:68]: Fixed

File: jmix-search/search/src/main/java/io/jmix/search/index/IndexSchemaManagementStrategy.java [hunk 19]
```diff
@@ -31,6 +31,12 @@ public enum IndexSchemaManagementStrategy {
      * Creates missing but skips existent indexes.
      */
     CREATE_ONLY("create-only"),
+
+    /**
+     * Creates missing but skips existent indexes. Updates index(updates mapping or configuration) it is possible.
+     */
+    CREATE_OR_CONFIGURATION_UPDATE("create-or-configuration-update"),
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/IndexSchemaManagementStrategy.java:38]: Fixed

File: jmix-search/search/src/main/java/io/jmix/search/index/IndexSchemaManagementStrategy.java [hunk 20]
```diff
@@ -73,4 +79,17 @@ public static IndexSchemaManagementStrategy getByKeyOrDefault(String key, IndexS
     public static IndexSchemaManagementStrategy getByKeyOrNull(String key) {
         return ENUM_MAP.get(key.toLowerCase());
     }
+
+    public boolean canUpdateConfiguration(){
```
[the remark this answers is not in the document]
Comment (fractal3000) [reply jmix-search/search/src/main/java/io/jmix/search/index/IndexSchemaManagementStrategy.java:83]: Fixed

**Comment by fractal3000** [comment]
Will be merged in https://github.com/jmix-framework/jmix/pull/3565

**Pull request closed without merging**